### ISSUE TITLE
benchmarks/distributed/ddp/benchmark.py raise error

### ISSUE THREAD

**Issue by zhiyuone** [pytorch/pytorch#50679]
## 🐛 Bug

benchmarks/distributed/ddp/benchmark.py running crash.

## To Reproduce

```bash
RANK=0 python3 pytorch-dist-benchmark.py --world-size 2 --master-addr 127.0.0.1 --master-port 23456
RANK=1 python3 pytorch-dist-benchmark.py --world-size 2 --master-addr 127.0.0.1 --master-port 23456
```

Crashes with the following error
```bash
Benchmark: resnet50 with batch size 32

                            sec/iter    ex/sec      sec/iter    ex/sec      sec/iter    ex/sec      sec/iter    ex/sec
Traceback (most recent call last):
  File "pytorch-dist-benchmark.py", line 320, in <module>
    main()
  File "pytorch-dist-benchmark.py", line 299, in main
    result = sweep(benchmark)
  File "pytorch-dist-benchmark.py", line 172, in sweep
    measurements = run_benchmark(benchmark, ranks, opts)
  File "pytorch-dist-benchmark.py", line 121, in run_benchmark
    return list(itertools.chain(*allgather_object(measurements)))
  File "pytorch-dist-benchmark.py", line 54, in allgather_object
    output.append(torch.load(buffer))
  File "/******/python3.7/lib/python3.7/site-packages/torch/serialization.py", line 587, in load
    with _open_zipfile_reader(opened_file) as opened_zipfile:
  File "/******/python3.7/lib/python3.7/site-packages/torch/serialization.py", line 242, in __init__
    super(_open_zipfile_reader, self).__init__(torch._C.PyTorchFileReader(name_or_buffer))
RuntimeError: [enforce fail at inline_container.cc:145] . PytorchStreamReader failed reading zip archive: invalid header or archive is corrupted
```

## Environment
```
PyTorch version: 1.7.1+cu110
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.3 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
Clang version: Could not collect
CMake version: version 3.5.1

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 455.45.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] numpy==1.17.0
[pip3] torch==1.7.1+cu110
[pip3] torch-struct==0.4
[pip3] torchaudio==0.7.2
[pip3] torchfile==0.1.0
[pip3] torchtext==0.8.1
[pip3] torchvision==0.8.2+cu110
[pip3] torchviz==0.0.1
[conda] Could not collect
```
## Additional context

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd

**Comment by rohan-varma** [comment]
Thanks for reporting this issue! I think we can use the newly available object-based APIs instead of the custom allgather here, I was able to repro the issue and get the benchmark running with this patch:

```
1	
diff --git a/benchmarks/distributed/ddp/benchmark.py b/benchmarks/distributed/ddp/benchmark.py
index 202ad3c5f5..3dcc5d2830 100644
--- a/benchmarks/distributed/ddp/benchmark.py
+++ b/benchmarks/distributed/ddp/benchmark.py
@@ -27,23 +27,9 @@ import torchvision
 
 
 def allgather_object(obj):
-    buffer = io.BytesIO()
-    torch.save(obj, buffer)
-    input_tensor = torch.ByteTensor(list(buffer.getvalue()))
-    input_length = torch.IntTensor([input_tensor.size(0)])
-    dist.all_reduce(input_length, op=dist.ReduceOp.MAX)
-    input_tensor.resize_(input_length[0])
-    output_tensors = [
-        torch.empty(input_tensor.size(), dtype=torch.uint8)
-        for _ in range(dist.get_world_size())
-    ]
-    dist.all_gather(output_tensors, input_tensor)
-    output = []
-    for tensor in output_tensors:
-        buffer = io.BytesIO(np.asarray(tensor).tobytes())
-        output.append(torch.load(buffer))
-    return output
-
+    out = [None for _ in range(dist.get_world_size())]
+    dist.all_gather_object(out, obj)
+    return out
 
 def allgather_run(cmd):
     proc = subprocess.run(shlex.split(cmd), capture_output=True)
```

Let us know if it works for your case and we can create a PR fixing this. Thanks!

**Comment by zhiyuone** [comment]
> Thanks for reporting this issue! I think we can use the newly available object-based APIs instead of the custom allgather here, I was able to repro the issue and get the benchmark running with this patch:
> 
> ```
> 1	
> diff --git a/benchmarks/distributed/ddp/benchmark.py b/benchmarks/distributed/ddp/benchmark.py
> index 202ad3c5f5..3dcc5d2830 100644
> --- a/benchmarks/distributed/ddp/benchmark.py
> +++ b/benchmarks/distributed/ddp/benchmark.py
> @@ -27,23 +27,9 @@ import torchvision
>  
>  
>  def allgather_object(obj):
> -    buffer = io.BytesIO()
> -    torch.save(obj, buffer)
> -    input_tensor = torch.ByteTensor(list(buffer.getvalue()))
> -    input_length = torch.IntTensor([input_tensor.size(0)])
> -    dist.all_reduce(input_length, op=dist.ReduceOp.MAX)
> -    input_tensor.resize_(input_length[0])
> -    output_tensors = [
> -        torch.empty(input_tensor.size(), dtype=torch.uint8)
> -        for _ in range(dist.get_world_size())
> -    ]
> -    dist.all_gather(output_tensors, input_tensor)
> -    output = []
> -    for tensor in output_tensors:
> -        buffer = io.BytesIO(np.asarray(tensor).tobytes())
> -        output.append(torch.load(buffer))
> -    return output
> -
> +    out = [None for _ in range(dist.get_world_size())]
> +    dist.all_gather_object(out, obj)
> +    return out
>  
>  def allgather_run(cmd):
>      proc = subprocess.run(shlex.split(cmd), capture_output=True)
> ```
> 
> Let us know if it works for your case and we can create a PR fixing this. Thanks!

sorry for so late to reply. It work for me.

**Comment by rohan-varma** [comment]
Fixed by #51095
